### PR TITLE
Inline selections - remaining stuff

### PR DESCRIPTION
**Pull request by JanKanis**
Finish the inline selection functionality and add proper tests.

I also took the opportunity to fold ConstraintDimension into the DimensionImpl classes.

### CODE REVIEW

File: transmart-core-db-tests/src/integration-test/groovy/org/transmartproject/db/multidimquery/QueryServiceSpec.groovy [hunk 1]
```diff
@@ -374,4 +380,56 @@ class QueryServiceSpec extends TransmartSpecification {
...
+        def result = multiDimService.retrieveClinicalData(constraint, accessLevelTestData.users[0]).asList()
+        def expectedObservations = hypercubeTestData.clinicalData.longitudinalClinicalFacts
+
+        then:
+        result.size() == expectedObservations.size()
+        result.every { it[DimensionImpl.STUDY] == testObservation.trialVisit.study }
+        result*.value.sort() == expectedObservations*.value.sort()
+    }
+
+    @Ignore("H2 does not support tuple comparisons with IN, which is used for subselections on visits, so this functionality will only work with Postgres or Oracle (Error: Subquery is not a single column query)")
```
Comment (ewelinagr) [inline transmart-core-db-tests/src/integration-test/groovy/org/transmartproject/db/multidimquery/QueryServiceSpec.groovy]: So maybe it is better to add test like this to transmart-core-db tests, which are running against postgres

File: transmart-core-db/src/main/groovy/org/transmartproject/db/multidimquery/query/HibernateCriteriaQueryBuilder.groovy [hunk 2]
```diff
@@ -434,33 +437,56 @@ class HibernateCriteriaQueryBuilder implements QueryBuilder<Criterion, DetachedC
...
+                // select trial visits from studies
+                def subQuery2 = DetachedCriteria.forClass(TrialVisit)
+                subQuery2.projection = Projections.property('id')
+                subQuery2.add(Subqueries.propertyIn('study', subQuery1))
+
+                // limit to the last set of trial visits
+                return Subqueries.propertyIn('trialVisit', subQuery2)
+
+//                return Subqueries.propertyIn(getAlias('trialVisit')+".study", subQuery)
+//                throw new QueryBuilderException("${constraint.constraintName} constraints for the study dimension are" +
```
Comment (ewelinagr) [inline transmart-core-db/src/main/groovy/org/transmartproject/db/multidimquery/query/HibernateCriteriaQueryBuilder.groovy]: How about the lines that are commented out, are they still needed?

Comment (JanKanis) [inline transmart-core-db/src/main/groovy/org/transmartproject/db/multidimquery/query/HibernateCriteriaQueryBuilder.groovy]: No, I removed them

File: transmart-core-db-tests/src/integration-test/groovy/org/transmartproject/db/multidimquery/QueryServiceSpec.groovy (hunk 1 above)
Comment (JanKanis) [inline transmart-core-db-tests/src/integration-test/groovy/org/transmartproject/db/multidimquery/QueryServiceSpec.groovy]: I'll do that too

**Pull request merged**